spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala

```diff
@@ -17,86 +17,95 @@
 package com.intel.analytics.bigdl.dataset
 
 import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
-import com.intel.analytics.bigdl.tensor.{DoubleType, FloatType, Storage, Tensor}
+import com.intel.analytics.bigdl.tensor.Tensor
+import org.apache.commons.lang3.SerializationUtils
 
 import scala.reflect.ClassTag
 
 /**
- * Sample, bundling input and target
- *
- * @param featureTensor
- * @param labelTensor
- * @tparam T
+ * Class that represents the features and labels of a data sample.
+ * @tparam T numeric type
  */
+abstract class Sample[T: ClassTag] extends Serializable {
+  /**
+   * Number of elements in features and labels.
+   * This function will be used in sorting samples.
+   * @return
+   */
+  def nElement(): (Int, Int)
 
-class Sample[T: ClassTag] (
-                            protected var featureTensor: Tensor[T],
-                            protected var labelTensor: Tensor[T])
-                          (implicit ev: TensorNumeric[T])
-  extends Serializable {
+  /**
+   * Copy other Sample's data to this Sample
+   * @param other Sample to be copied.
+   * @return this
+   */
+  def copy(other: Sample[T]): this.type
 
-  def this()(implicit ev: TensorNumeric[T]) = this(Tensor[T](), Tensor[T]())
+  /**
+   * Number of features in this Sample
+   * @return number of features
+   */
+  def numFeature(): Int
```

Comment:
For the multi-inputs sample, is the number of inputs. I use this to create a array buffer.